nexus/db-model/src/project.rs

```diff
@@ -40,6 +48,61 @@ impl From<Project> for views::Project {
     }
 }
 
+impl DatastoreCollectionConfig<Instance> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = instance::dsl::project_id;
+}
+
+impl DatastoreCollectionConfig<Disk> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = disk::dsl::project_id;
+}
+
+impl DatastoreCollectionConfig<Image> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = image::dsl::project_id;
+}
+
+impl DatastoreCollectionConfig<Snapshot> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = snapshot::dsl::project_id;
+}
+
+impl DatastoreCollectionConfig<Vpc> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = vpc::dsl::project_id;
+}
+
+// NOTE: "IpPoolRange" also contains a reference to "project_id", but
+// ranges should only exist within IP Pools.
+impl DatastoreCollectionConfig<IpPool> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = ip_pool::dsl::project_id;
+}
+
+// TODO(https://github.com/oxidecomputer/omicron/issues/1482): Not yet utilized,
+// but needed for project deletion safety.
+// TODO(https://github.com/oxidecomputer/omicron/issues/1334): Cannot be
+// utilized until floating IPs are implemented.
+impl DatastoreCollectionConfig<ExternalIp> for Project {
+    type CollectionId = Uuid;
+    type GenerationNumberColumn = project::dsl::rcgen;
+    type CollectionTimeDeletedColumn = project::dsl::time_deleted;
+    type CollectionIdColumn = external_ip::dsl::project_id;
+}
+
```

Comment:
Ehhh that's kinda a blurry line. We do sometimes (and I wouldn't be opposed to having one!), but generally I've favored "feel the initial pain of writing it out a lot" rather than "be as brief as possible", until the benefit of the macro is immediately obvious.

I think right now we have a few attribute macros for DB queries and metrics-querying support, and a handful of macros for authz stuff? `nexus/db-macros` and `nexus/authz-macros` are examples. 

But yeah, that question of readability is a bit subjective -- do you think this would benefit from a macro? My opinion *had* been that this `impl` is specifying a handful of fields we'd want to supply through a macro anyway, so I figured a macro would just obscure the intent, but I'm curious if you have an example in mind where this might seem clearer.

Reply:
Oh I was thinking of a private macro here, not a public API. Though the primary benefit I've found is if we expect many more cases and it enhances code clarity (e.g. [this is a place where macros make life better](https://github.com/camino-rs/camino/blob/61c16e917ac365b0ebecf5559f69a0c75ec9236b/src/lib.rs#L2796-L2806)). I don't think this crosses that bar.